crud/program/src/instruction.rs

```diff
@@ -0,0 +1,135 @@
+//! Program instructions
+
+use crate::{state::Data, *};
+use borsh::{BorshDeserialize, BorshSerialize};
+use solana_program::{
+    instruction::{AccountMeta, Instruction},
+    pubkey::Pubkey,
+    sysvar,
+};
+
+/// Instructions supported by the Feature Proposal program
+#[derive(Clone, Debug, BorshSerialize, BorshDeserialize, PartialEq)]
+pub enum CrudInstruction {
+    /// Create a new document
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[writeable, signer]` Data account, must be uninitialized
+    /// 1. `[]` Document owner
+    /// 2. `[]` Rent sysvar, to check for rent exemption
+    Create {
+        /// Data to be filled into the account
+        data: Data,
+    },
+
+    /// Update the provided document account
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    /// 0. `[writeable]` Document account, must be previously initialized (version != 0)
+    /// 1. `[signer]` Current owner of the document
+    Update {
+        /// Data to replace the existing document data
+        data: Data,
```

Comment:
Supporting partial writes would be really nice

Reply:
Do you have a neat way in mind to do this?  We could come up with a derivable traits on `Data` which allows for merging

Reply:
I was just thinking of `Write { offset: u32, data: Vec<u32> }`, nothing fancy

Reply:
Done in d57870f